### PR TITLE
chore: Description for no_inline attribute

### PR DESCRIPTION
**Pull request by MikaelMayer**
The attribute `{:no_inline}` was never documented, so I tried to understand how it works and wrote about it.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/Dafny/DafnyOptions.cs [hunk 1]
```diff
@@ -827,7 +827,25 @@ public enum IncludesModes {
...
+      Hence, sometimes, if sound to do so (e.g. no `{:opaque}`),
+      Dafny will inline the function, resulting in, for example,
+      `assert 6 % 2 == 0`.
+      
+      Adding the attribute `{:no_inline}` to a function will result
+      in the Dafny verifier to never inline the function in such
+      a function call, but unless the function is `{:opaque}`, its
+      definition will still be available.
+      
+      This trick can be helpful if for a huge conjunct predicate `P`,
```
Comment (keyboardDrummer) [inline Source/Dafny/DafnyOptions.cs:843]: I don't understand this example. How do P, P(x) and f(x) relate?

File: Source/Dafny/DafnyOptions.cs [hunk 2]
```diff
@@ -827,7 +827,25 @@ public enum IncludesModes {
...
+      assuming that `P(x)` already hold, if we don't want `P`
+      to be opaque, and we `assert f(x)` again. Inlining might result
```
Comment (MikaelMayer) [inline Source/Dafny/DafnyOptions.cs:845]:
```suggestion
      to be opaque, and we `assert f(x)` again. Inlining might result
```
```suggestion
      to be opaque, and we `assert P(x)` again. Inlining might result
```

**Review by MikaelMayer — COMMENTED** [review]
Fixing notations

File: Source/Dafny/DafnyOptions.cs (hunk 1 above)
Comment (MikaelMayer) [reply Source/Dafny/DafnyOptions.cs:843]: > I don't understand this example. How do P, P(x) and f(x) relate?

Fixed. It should have been all P

File: Source/Dafny/DafnyOptions.cs [hunk 3]
```diff
@@ -827,7 +827,25 @@ public enum IncludesModes {
       TODO
 
     {:no_inline}
-      TODO
+      When predicates such as `function P(x: int) { x % 2 == 0 }`
```
Comment (MikaelMayer) [inline Source/Dafny/DafnyOptions.cs:830]:
```suggestion
      When predicates such as `predicate P(x: int) { x % 2 == 0 }`
```

File: Source/Dafny/DafnyOptions.cs [hunk 4]
```diff
@@ -827,7 +827,25 @@ public enum IncludesModes {
       TODO
 
     {:no_inline}
-      TODO
+      When predicates such as `predicate P(x: int) { x % 2 == 0 }`
+      are used in assertions like `assert P(6);`, Dafny
+      will by default try to figure out if it can split the call
+      into multiple assertions that are easier for the verifier.
+      Hence, sometimes, if sound to do so (e.g. no `{:opaque}`),
+      Dafny will inline the function, resulting in, for example,
```
Comment (MikaelMayer) [inline Source/Dafny/DafnyOptions.cs:835]:
```suggestion
      Dafny will inline the predicate, resulting in, for example,
```

File: Source/Dafny/DafnyOptions.cs [hunk 5]
```diff
@@ -858,7 +858,25 @@ public enum IncludesModes {
       TODO
 
     {:no_inline}
-      TODO
+      When predicates such as `predicate P(x: int) { x % 2 == 0 }`
+      are used in assertions like `assert P(6);`, Dafny
+      will by default try to figure out if it can split the call
+      into multiple assertions that are easier for the verifier.
+      Hence, sometimes, if sound to do so (e.g. no `{:opaque}`),
```
Comment (cpitclaudel) [inline Source/Dafny/DafnyOptions.cs:865]:
```suggestion
      Hence, sometimes, if allowed to do so (e.g. no `{:opaque}`),
```

File: Source/Dafny/DafnyOptions.cs [hunk 6]
```diff
@@ -858,7 +858,25 @@ public enum IncludesModes {
...
+      Dafny will inline the predicate, resulting in, for example,
+      `assert 6 % 2 == 0`.
+      
+      Adding the attribute `{:no_inline}` to a function will result
+      in the Dafny verifier to never inline the function in such
+      a function call, but unless the function is `{:opaque}`, its
+      definition will still be available.
```
Comment (cpitclaudel) [inline Source/Dafny/DafnyOptions.cs:872]:
```suggestion
      Adding the attribute `{:no_inline}` to a function will prevent
      the Dafny verifier from inlining it, but unless the function is
      `{:opaque}` its definition will still be available.
```

File: Source/Dafny/DafnyOptions.cs [hunk 7]
```diff
@@ -858,7 +858,25 @@ public enum IncludesModes {
...
+      
+      This trick can be helpful if for a huge conjunct predicate `P`,
+      assuming that `P(x)` already hold, if we don't want `P`
+      to be opaque, and we `assert P(x)` again. Inlining might result
+      in performance issues because it will have to infer every single
+      conjunct. Adding `{:no_inline}` to the predicate can result
+      in such cases in the verifier to be faster.
```
Comment (cpitclaudel) [inline Source/Dafny/DafnyOptions.cs:879]: The grammar need a bit of touch up here (extra `if` in `if we don't want …`) and `can result in … being`.

File: Source/Dafny/DafnyOptions.cs [hunk 8]
```diff
@@ -858,7 +858,25 @@ public enum IncludesModes {
...
+      Hence, sometimes, if sound to do so (e.g. no `{:opaque}`),
+      Dafny will inline the predicate, resulting in, for example,
+      `assert 6 % 2 == 0`.
+      
+      Adding the attribute `{:no_inline}` to a function will result
+      in the Dafny verifier to never inline the function in such
+      a function call, but unless the function is `{:opaque}`, its
+      definition will still be available.
+      
+      This trick can be helpful if for a huge conjunct predicate `P`,
```
Comment (MikaelMayer) [inline Source/Dafny/DafnyOptions.cs:874]:
```suggestion
      This trick can be helpful, for a huge conjunct predicate `P`,
```

File: Source/Dafny/DafnyOptions.cs (hunk 7 above)
Comment (MikaelMayer) [inline Source/Dafny/DafnyOptions.cs:879]:
```suggestion
      in such cases in the verifier being faster.
```

**Review by cpitclaudel — APPROVED** [review]
👍

**Pull request merged**